### PR TITLE
 Treat invokespecial symbolic reference as superclass in method search

### PR DESCRIPTION
**Pull request by theresa-m**
From JVMS: If the symbolic reference names a class (not an interface), then
that class is a superclass of the current class.
    
Previously the algorithm for finding the invoked method for invokespecial
assumes that the class named through the bytecode symbolic reference is the
current class when it is actually the superclass of the current class. By
doing this it skips searching direct interfaces of the referenced class for
default interface methods.

Fixes: https://github.com/eclipse-openj9/openj9/issues/18705

### CODE REVIEW

**Comment by tajila** [comment]
@ChengJin01 Please review

**Comment by ChengJin01** [comment]
1) If I understand correctly in the code, the class hierarchy chain of the failing test is something like:
Class3 --> Class 1 ---> Class2 ---> Interface1 ---> Interface2
where there are two non-abstract methods `m()` (with the same naming and different return value) placed in  `Interface1` and `Interface2` separately.
```
public interface Interface1 extends Interface2
  minor version: 0
  major version: 52
  flags: (0x0601) ACC_PUBLIC, ACC_INTERFACE, ACC_ABSTRACT

  public default java.lang.Integer m();
    descriptor: ()Ljava/lang/Integer;
    flags: (0x0001) ACC_PUBLIC
    Code:
      stack=3, locals=1, args_size=1
         0: new           #1                  // class java/lang/Integer
         3: dup
         4: iconst_1  <---------
         5: invokespecial #2                  // Method java/lang/Integer."<init>":(I)V
         8: areturn

public interface Interface2
  minor version: 0
  major version: 52
  flags: (0x0601) ACC_PUBLIC, ACC_INTERFACE, ACC_ABSTRACT

  public default java.lang.Integer m();
    descriptor: ()Ljava/lang/Integer;
    flags: (0x0001) ACC_PUBLIC
    Code:
      stack=3, locals=1, args_size=1
         0: new           #1                  // class java/lang/Integer
         3: dup
         4: iconst_2 <-------
         5: invokespecial #2                  // Method java/lang/Integer."<init>":(I)V
         8: areturn
```
The question with this situation is how to determine which `m()` (from the interfaces) should be selected based on the class hierarchy above against the JVM Spec.

2. I roughly went though the JVM Spec at https://docs.oracle.com/javase/specs/jvms/se11/jvms11.pdf and it seems still unclear to me how the selection logic is determined given both `invokevirtual` and `invokespecial` follows the same rules for `invokeinterface` as follows:
```
invokevirtual:
It is possible that the symbolic reference of an invokevirtual
instruction resolves to an interface method. In this case, it is
possible that there is no overriding method in the class hierarchy,
but that a non-abstract interface method matches the resolved
method's descriptor. The selection logic matches such a method,
using the same rules as for invokeinterface.

invokespecial:
The invokespecial instruction handles invocation of a non-
abstract interface method, referenced either via a direct
superinterface or via a superclass. In these cases, the rules for
selection are essentially the same as those for invokeinterface
(except that the search starts from a different class).

invokeinterface: 
The selection logic allows a non-abstract method declared in
a superinterface to be selected. Methods in interfaces are only
considered if there is no matching method in the class hierarchy.
In the event that there are two non-abstract methods in the
superinterface hierarchy, with neither more specific than the other,
an error occurs; there is no attempt to disambiguate (for example,
one may be the referenced method and one may be unrelated, but
we do not prefer the referenced method). On the other hand, if there
are many abstract methods but only one non-abstract method,
the non-abstract method is selected (unless an abstract method
is more specific)
```

Please note the explanation of `invokeinterface` above as to two non-abstract methods:
```
Methods in interfaces are only considered if there is no matching method in the class hierarchy.
In the event that there are two non-abstract methods in the superinterface hierarchy, with neither
more specific than the other, an error occurs; <--------
there is no attempt to disambiguate...
```
My understanding is that an error should be thrown out by JVM when the ambiguous case happens instead of
simply selecting non-abstract method from the closest superinterface.

That being said, we probably need more concrete clarifications from Oracle in such case before moving forward to fix the issue.

File: runtime/util/resolvehelp.c [hunk 1]
```diff
@@ -46,9 +46,16 @@ getMethodForSpecialSend(J9VMThread *vmStruct, J9Class *currentClass, J9Class *re
...
 		UDATA superDepth = J9CLASS_DEPTH(methodClass);
-		BOOLEAN isInterfaceMethod = J9AccInterface == (methodClass->romClass->modifiers & J9AccInterface);
 
-		if ((isInterfaceMethod || ((currentDepth > superDepth) && (currentClass->superclasses[superDepth] == methodClass)))
+		/* Exclude default interface methods.
+		 * JVM spec: The invokespecial instruction handles invocation of a
+		 * nonabstract interface method... the rules for selection are
+		 * essentially the same as those for invokeinterface.
+		 */
+		BOOLEAN isAbstractInterfaceMethod = (J9AccInterface == (methodClass->romClass->modifiers & J9AccInterface))
```
Comment (ChengJin01) [inline runtime/util/resolvehelp.c:55]: It'd better to use `J9_ARE_ALL_BITS_SET` at this point.

File: runtime/util/resolvehelp.c [hunk 2]
```diff
@@ -46,9 +46,16 @@ getMethodForSpecialSend(J9VMThread *vmStruct, J9Class *currentClass, J9Class *re
...
+								&& (J9_ROM_METHOD_FROM_RAM_METHOD(method)->modifiers & J9AccAbstract);
+
+		if ((isAbstractInterfaceMethod || ((currentDepth > superDepth) && (currentClass->superclasses[superDepth] == methodClass)))
```
Comment (ChengJin01) [inline runtime/util/resolvehelp.c:58]: It should be nice to adjust the indentation for the `if` statement. e.g.
```
if ((isAbstractInterfaceMethod 
   || ((currentDepth > superDepth) && (currentClass->superclasses[superDepth] == methodClass)))
   && (J9_ARE_NO_BITS_SET(resolvedClass->romClass->modifiers, J9AccInterface) 
   && J9_ARE_NO_BITS_SET(currentClass->romClass->modifiers, J9AccInterface))
 ) {
```

**Review by ChengJin01 — CHANGES_REQUESTED** [review]

**Comment by theresa-m** [comment]
Thanks for looking this over @ChengJin01 and pointing out the additional details from invokeinterface.

> That being said, we probably need more concrete clarifications from Oracle in such case before moving forward to fix the issue.

What do you suggest to achieve this?

**Comment by ThanHenderson** [comment]
I think an important thing to figure out here is what `...with neither more specific than the other...` means.

**Comment by ChengJin01** [comment]
It definitely ends up with error in the case of the multiple inheritance as @ThanHenderson mentioned above.

Now the question whether it is acceptable/reasonable to stop at the closest interface (`Interface 1`) for the concrete method in the class hierarchy chain when searching from bottom to top. based on the explanation of JVM Spec. If so, it seems like `m()` in `Interface2` never got a chance to be invoked. @tajila will double-check this situation to see whether we need to contact Oracle for this ambiguous case.

**Comment by theresa-m** [comment]
Thanks @tajila 

For what its worth when talking about the class relationships for the example neither diagrams are exactly correct, Class1 implements Interface1:

```
Class3->Class1---->Class2
           |         |
      Interface1->Interface2
```
I usually interpret "more specific" for method resolution as related to the method signature and for this case they are the same.

**Comment by ChengJin01** [comment]
> For what its worth when talking about the class relationships for the issue neither diagrams are exactly correct, Class1 implements Interface1 and Class2 implements Interface2 ....

Fundamentally, there might be no big difference as compared to the previous example above. I think the JVM Spec doesn't make it clear at this point as to whether a shortest searching depth in the class hierarchy is always true for such corner cases, assuming there is no concrete method in the superclasses that matches the specified method.

**Comment by tajila** [comment]
My understanding is that in this case since we are dealing with an instance interface method the maximally specific method should be selected

```
Otherwise, if there is exactly one maximally-specific method ([§5.4.3.3](https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-5.html#jvms-5.4.3.3)) in the superinterfaces of C that matches the resolved method's name and descriptor and is not abstract, then it is the method to be invoked.
```


Given the following 
```
 Where the method is declared in interface I, there exists no other maximally-specific superinterface method of C with the
 specified name and descriptor that is declared in a subinterface of I. 
```

This would imply that from the perspective of an implementing class (in this case Class3), if there exist two interfaces (`Interface1` and `Interface2`) that define an instance method with the same method signature (`m()`), and if one of the interfaces is a super-type of the other, then the super-type is not the "maximally-specific superinterface method". 

This suggests that when dealing with an interface hierarchy where an instance interface method as the same signature throughout that hierarchy, and the instance interface method is also define throughout the hierarchy, the interface method at the bottom of the hierarchy is selected.

**Comment by theresa-m** [comment]
Thanks for clarifying. That looks to be the behavior already in `javaLookupMethod` ([called from `getMethodForSpecialSend`](https://github.com/eclipse-openj9/openj9/blob/master/runtime/vm/resolvesupport.cpp#L1352)).

Where it goes wrong is `getMethodForSpecialSend` which refers to a ["super send" situation](https://github.com/eclipse-openj9/openj9/blob/master/runtime/util/resolvehelp.c#L36). I'm not finding reference to that in the spec but perhaps CMVC 170457 would shed some light if anyone has access to that. My idea with this pull request is to avoid it for default interface methods since invokespecial notes this handling should be the same as invokevirtual, which rely's on `javaLookupMethod`.

**Comment by tajila** [comment]
> Where it goes wrong is getMethodForSpecialSend which refers to a ["super send" situation](https://github.com/eclipse-openj9/openj9/blob/master/runtime/util/resolvehelp.c#L36). I'm not finding reference to that in the spec

"Super send" is smalltalk nomenclature for invoking a supertype's method. Small talk had the notion of "sending" messages and "receiving" messages, which roughly translates to `invoke*` and `receivers` in Java.

**Comment by theresa-m** [comment]
Thanks everyone for the help interpreting the spec.

Upon further inspection I see that the interface search starts from the [superclass of the current class](https://github.com/eclipse-openj9/openj9/blob/master/runtime/util/resolvehelp.c#L61).

In this case the current class is class1 since that is where the method is called from. The search then starts at Class2 which will miss Interface1 altogether.

I will update my changes shortly, and also the super send language since thats pretty out of date at this point.

**Comment by ChengJin01** [comment]
> ...  In this case the current class is class1 since that is where the method is called from. The search then starts at Class2 which will miss Interface1 altogether.

Base on the test code listed at https://github.com/eclipse-openj9/openj9/issues/18705
```
public class Class1 extends Class2 implements Interface1 {
    public Class1() {
    }
    public static Integer test() {
        Class3 var1 = new Class3(); 
        return var1.m(); <------ 
    }
}
```
it seems to me that the the receiver class for the method `m()` being invoked should be `Class 3` rather than `Class 1`. If so, the searching from the bottom of the interface hierarchy should still end up with `interface1::m()` given there is no concrete method in the superclasses.

**Comment by theresa-m** [comment]
My thought was based on from the spec:

> The difference between the invokespecial instruction and the
invokevirtual instruction (§invokevirtual) is that invokevirtual
invokes a method based on the class of the object. The
invokespecial instruction is used to directly invoke instance
initialization methods (§2.9.1) as well as methods of the current
class and its supertypes.

that for this case the class of the object is Class3 and the current class is Class1. But I need to confirm my understanding of current class.

**Comment by theresa-m** [comment]
Under invokespecial verification I see the following: 
> starting with the current class (the class performing invokespecial).

Class1 is that class that performs invokespecial.

**Comment by theresa-m** [comment]
Okay my new theory is that Class3 is the current class. Because despite what I referenced above being true when determining the 'c' in the invokespecial algorithm this line also exists:
> • If the symbolic reference names a class (not an interface), then that class is a superclass of the current class.

```
#3 = Methodref          #9.#17         // Class1.m:()Ljava/lang/Integer;
#9 = Class              #11            // Class1
#17 = NameAndType        #14:#16        // m:()Ljava/lang/Integer;
...
9: invokespecial #3                  // Method m:()Ljava/lang/Integer;
```

From the above bytecode snippet you can see that the symbolic reference names Class1 which is a class not an interface.
Then Class1 is a superclass of the current class which has to be Class3.

**Comment by ChengJin01** [comment]
I just found out more specific explanation on `the maximally-specific superinterface method` at the `
5.4.3.3 Method Resolution` section at https://docs.oracle.com/javase/specs/jvms/se21/jvms21.pdf as follows in case we ignored anything with the terminology against the JVM Spec:

5.4.3.3 Method Resolution
...
1. If C is an interface, method resolution throws an IncompatibleClassChangeError
2. Otherwise, method resolution attempts to locate the referenced method in C and its superclasses
3. Otherwise, method resolution attempts to locate the referenced method in the superinterfaces of the specified class C:
• If the **maximally-specific** superinterface methods of C for the name and
descriptor specified by the method reference include exactly one method that
does not have its ACC_ABSTRACT flag set, then this method is chosen and
method lookup succeeds.
...
A **maximally-specific** superinterface method of a class or interface C for a particular
method name and descriptor is any method for which all of the following are true:
• The method is declared in a superinterface (**direct** or indirect) of C. <----------
• The method is declared with the specified name and descriptor.
• The method has neither its ACC_PRIVATE flag nor its ACC_STATIC flag set.
• Where the method is declared in interface I, there exists no other maximally-
specific superinterface method of C with the specified name and descriptor that
is declared in a subinterface of I.

So it turns out that the conclusion is the same as before: `Interface_1::m()` should be selected.

**Comment by theresa-m** [comment]
Thanks @ChengJin01 I just pushed a new solution which corrects which 'c' we are searching. Let me know what you think.

**Comment by ChengJin01** [comment]
> Thanks @ChengJin01 I just pushed a new solution which corrects which 'c' we are searching. Let me know what you think.

@theresa-m, please double-check to see whether all external & internal test suites passes with your changes to avoid any corner cases & conflicts in the existing implementations.

**Review by ChengJin01 — APPROVED** [review]

**Comment by theresa-m** [comment]
@TobiAjila can you take a look as well?

**Comment by theresa-m** [comment]
The push was to rebase this change with the master branch.

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
jenkins test sanity win jdk8

**Comment by tajila** [comment]
jenkins test sanity plinux jdk21
